libopenage/gamestate/score.h

```diff
@@ -0,0 +1,139 @@
+// Copyright 2017-2017 the openage authors. See copying.md for legal info.
+
+#pragma once
+
+#include <functional>
+
+#include "resource.h"
+
+
+namespace openage {
+
+class Player;
+class Team;
+
+/**
+ * The categories of sub-scores that sum to a player's score.
+ */
+enum class score_category {
+	/** 20% of units killed cost */
+	military,
+	/** 20% of alive units cost and 10% of resources */
+	economy,
+	/** 20% of researched technologies and 10 for each 1% of map explored*/
+	technology,
+	/** 20% of Castles and Wonders cost */
+	society,
+	SCORE_CATEGORY_COUNT
+};
+
+/**
+ * Keeps track of a score and all the sub-scores
+ */
+class Score {
+public:
+
+	Score();
+
+	void add_score(const score_category cat, double value);
+	void add_score(const score_category cat, int value);
+
+	void remove_score(const score_category cat, double value);
+	void remove_score(const score_category cat, int value);
+
+	/**
+	 * Updates map exploration precentance based sub-scores
+	 */
+	void update_map_explored(double progress);
+
+	/**
+	 * Updates resource based sub-scores
+	 */
+	void update_resources(const ResourceBundle & resources);
+
+	/**
+	 * Calculates the total score from the sub-scores.
+	 * TODO update gui here
+	 */
+	virtual void update_score();
+
+	// Getters
+
+	int getScore(const score_category cat) const { return score[(int) cat]; }
+	int getScore(const int index) const { return score[index]; }
+
+	int getScoreTotal() const { return score_total; }
+
+protected:
+
+	int score[(int) score_category::SCORE_CATEGORY_COUNT];
+
+	// generated values
+
+	int score_total;
+
+private:
+
+	/** Used by update_map_explored. */
+	int score_exploration;
+
+	/** Used by update_resources. */
+	int score_resources;
+};
+
+
+/**
+ * The score of a player
+ */
+class ScorePlayer : public Score {
```

Comment:
That makes sense. I'd still say we should focus on human-compatibility though, although the autocomplete argument is a good one, then again, how often do we need to complete those classes? :)